test/distributed/fsdp/test_fsdp_misc.py

```diff
@@ -86,6 +87,87 @@ def forward(self, x):
         # https://github.com/pytorch/pytorch/issues/83107 and
         # https://github.com/pytorch/pytorch/issues/83129
 
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_not_all_outputs_used_in_loss(self):
+
+        class MyModule(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.lin1 = nn.Linear(4, 4)
+                self.lin2 = nn.Linear(4, 4)
+
+            def forward(self, x):
+                a = self.lin1(x)
+                b = self.lin2(x)
+                return (a, b)
+
+        def _check_resharded(fsdp_module):
+            for param in fsdp_module.params:
+                if param._is_sharded:
+                    full_param = param._full_param_padded
+                    self.assertEqual(full_param.storage().size(), 0)
+
+                self.assertEqual(
+                    param.data.data_ptr(),
+                    param._local_shard.data_ptr()
+                )
+
+        def _check_equal(local, fsdp):
+            with FSDP.summon_full_params(fsdp):
+                for p1, p2 in zip(fsdp.parameters(), local.parameters()):
+                    torch.testing.assert_allclose(p1, p2)
+
+        for sharding_strategy in [
+            ShardingStrategy.FULL_SHARD,
+            ShardingStrategy.SHARD_GRAD_OP,
+            ShardingStrategy.NO_SHARD
+        ]:
+            with self.subTest(sharding_strategy=sharding_strategy):
```

Comment:
nit: Do you want to make this use `self.run_subtests()`? It would require factoring this test logic into its own function.

```
self.run_subtests(
    {"sharding_strategy": [..., ..., ...]},
    <test_name>,
    <test_args>,
    <test_kwargs>,
)
```

Reply:
Filed issue as follow up work: https://github.com/pytorch/pytorch/issues/84071